File: malai-core/org.malai.ts/src-core/fsm/SubFSMTransition.ts
/*
 * This file is part of Malai.
 * Copyright (c) 2009-2018 Arnaud BLOUIN
 * Malai is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * Malai is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

import {Transition} from "./Transition";
import {FSM} from "./FSM";
import {FSMHandler} from "./FSMHandler";
import {OutputState} from "./OutputState";
import {InputState} from "./InputState";
import {Optional} from "../../src/util/Optional";
import {TerminalState} from "./TerminalState";
import {CancellingState} from "./CancellingState";
import {OutputStateImpl} from "./OutputStateImpl";

export class SubFSMTransition<E> extends Transition<E> {
    private readonly subFSM: FSM<E>;

    private readonly subFSMHandler: FSMHandler;

    public constructor(srcState: OutputState<E>, tgtState: InputState<E>, fsm: FSM<E>) {
        super(srcState, tgtState);
        this.subFSM = fsm;
        this.subFSM.setInner(true);
        this.subFSMHandler = new class implements FSMHandler {
            protected _parent: SubFSMTransition<E>;

            constructor(parentFSM: SubFSMTransition<E>) {
                this._parent = parentFSM;
            }

            public fsmStarts(): void {
                this._parent.src.exit();
            }

            public fsmUpdates(): void {
                this._parent.src.getFSM().setCurrentState(this._parent.subFSM.getCurrentState());
                this._parent.src.getFSM().onUpdating();
            }

            public fsmStops(): void {
                this._parent.action(undefined);
                this._parent.subFSM.removeHandler(this._parent.subFSMHandler);
                this._parent.src.getFSM().setCurrentSubFSM(undefined);
                if (this._parent.tgt instanceof  TerminalState) {
                    this._parent.tgt.enter();
                    return;
                }
                if (this._parent.tgt instanceof CancellingState) {
                    this.fsmCancels();
                    return;
                }
                if (this._parent.tgt instanceof OutputStateImpl) {
                    this._parent.src.getFSM().setCurrentState(this._parent.tgt);
                    this._parent.tgt.enter();
                }
            }

            public fsmCancels(): void {
                this._parent.subFSM.removeHandler(this._parent.subFSMHandler);
                this._parent.src.getFSM().setCurrentSubFSM(undefined);
                this._parent.src.getFSM().onCancelling();
            }
        }(this);
    }

    /**
     * @param {*} event
     */
    public execute(event: E): Optional<InputState<E>> {
        if (this.isGuardOK(event)) {
            this.src.getFSM().stopCurrentTimeout();
            const transition: Optional<Transition<E>> = this.findTransition(event);
            if (transition.isPresent()) {
                this.subFSM.addHandler(this.subFSMHandler);
                this.src.getFSM().setCurrentSubFSM(this.subFSM);
                this.subFSM.process(event);
                return Optional.ofNullable(transition.get()).map(t => t.tgt);
            }
        }
        return Optional.empty<InputState<E>>();
    }

    /**
     *
     * @param {*} event
     * @return {boolean}
     */
    public accept(event: E): boolean {
        return this.findTransition(event).isPresent();
    }

    /**
     *
     * @param {*} event
     * @return {boolean}
     */
    public isGuardOK(event: E): boolean {
        return this.findTransition(event).filter(tr => tr.isGuardOK(event)).isPresent();
    }

    private findTransition(event: E): Optional<Transition<E>> {
        return Optional.ofNullable(this.subFSM.initState.getTransitions().find(tr => tr.accept(event)));
    }

    /**
     *
     * @return {*[]}
     */
    public getAcceptedEvents(): Set<String> {
        return this.subFSM.initState.getTransitions().map(tr => tr.getAcceptedEvents()).reduce((a, b) => new Set([...a, ...b]));
    }
}

File: src/main/java/io/github/interacto/fsm/SubFSMTransition.java
/*
 * Interacto
 * Copyright (C) 2020 Arnaud Blouin
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package io.github.interacto.fsm;

import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * A transition that refers to another FSM.
 * Entering this transition starts the underlying sub-FSM.
 * To leave the transition, the sub-FSM must end.
 * @param <E0> The type of events the FSM processes.
 */
public class SubFSMTransition<E0> extends Transition<E0, E0> {
	private final FSM<E0> subFSM;
	private final FSMHandler subFSMHandler;

	/**
	 * Creates the transition.
	 * @param srcState The source state of the transition.
	 * @param tgtState The output state of the transition.
	 * @param fsm The inner FSM that composes the transition.
	 * @throws IllegalArgumentException If one of the states is null.
	 */
	public SubFSMTransition(final OutputState<E0> srcState, final InputState<E0> tgtState, final FSM<E0> fsm) {
		super(srcState, tgtState);

		if(fsm == null) {
			throw new IllegalArgumentException("sub fsm cannot be null");
		}

		subFSM = fsm;
		subFSM.setInner(true);
		subFSMHandler = new FSMHandler() {
			@Override
			public void fsmStarts() throws CancelFSMException {
				src.exit();
			}

			@Override
			public void fsmUpdates() throws CancelFSMException {
				src.getFSM().setCurrentState(subFSM.getCurrentState());
				src.getFSM().onUpdating();
			}

			@Override
			public void fsmStops() throws CancelFSMException {
				action(null);
				subFSM.removeHandler(subFSMHandler);
				src.getFSM().currentSubFSM = null;
				if(tgt instanceof TerminalState) {
					tgt.enter();
					return;
				}
				if(tgt instanceof CancellingState) {
					fsmCancels();
					return;
				}
				if(tgt instanceof OutputState) {
					src.getFSM().setCurrentState((OutputState<E0>) tgt);
					tgt.enter();
				}
			}

			@Override
			public void fsmCancels() {
				subFSM.removeHandler(subFSMHandler);
				src.getFSM().currentSubFSM = null;
				src.getFSM().onCancelling();
			}
		};
	}

	@Override
	public Optional<InputState<E0>> execute(final E0 event) {
		final Optional<Transition<E0, E0>> transition = findTransition(event);

		if(transition.isPresent()) {
			src.getFSM().stopCurrentTimeout();
			subFSM.addHandler(subFSMHandler);
			src.getFSM().currentSubFSM = subFSM;
			subFSM.process(event);
			return Optional.of(transition.get().tgt);
		}

		return Optional.empty();
	}

	@Override
	protected E0 accept(final E0 event) {
		return findTransition(event).isEmpty() ? null : event;
	}

	@Override
	protected boolean isGuardOK(final E0 event) {
		return findTransition(event)
			.filter(tr -> tr.isGuardOK(event))
			.isPresent();
	}

	private Optional<Transition<E0, E0>> findTransition(final E0 event) {
		return subFSM.initState.transitions
			.stream()
			.filter(tr -> tr.accept(event) != null)
			.map(tr -> (Transition<E0, E0>) tr)
			.findFirst();
	}

	@Override
	public Set<Object> getAcceptedEvents() {
		return subFSM.initState
			.getTransitions()
			.stream()
			.map(tr -> tr.getAcceptedEvents())
			.flatMap(s -> s.stream())
			.collect(Collectors.toSet());
	}

	@Override
	public void uninstall() {
		subFSM.uninstall();
	}
}
